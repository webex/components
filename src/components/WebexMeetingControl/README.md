# Webex Meeting Control Component

Webex Meeting Control component displays a button that allows for control over a meeting by performing a specific action.

<p align="center">
  <img src="./WebexMeetingControl.png" alt="Default Webex Meeting Control" />
</p>

## Preview

To see all the different possible states of the Webex Meeting Control component, you can run our Storybook:

```shell
  npm start
```

## Embed

1. Create a component adapter from which the data will be retrieved (See [adapters](../../adapters)). For instance:

    ```js
    const jsonAdapter = new WebexJSONAdapter(jsonData);
    ```

2. Create a component instance by passing the control _type_ as a string and
enclose it within [a data provider](../WebexDataProvider/WebexDataProvider.js)
that takes the [component data adapter](../../adapters/WebexJSONAdapter.js) that we created previously.
A Webex Meeting Control must also be enclosed by a [Webex Meeting Controls](./WebexMeetingControls.js) component that contains the
meeting context for the control

    ```js
    <WebexDataProvider adapter={jsonAdapter}>
      <WebexMeetingControls meetingID="meetingID">
        <WebexMeetingControl type="mute-audio" />
      </WebexMeetingControls>
    </WebexDataProvider>
    ```

The component knows how to manage its data. If anything changes in the data source that the adapter manages, the component will also update on its own.
