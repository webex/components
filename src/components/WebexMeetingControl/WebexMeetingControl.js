import React from 'react';
import PropTypes from 'prop-types';
import {Button, Icon} from '@momentum-ui/react';
import {MeetingControlState} from '@webex/component-adapter-interfaces';

import {useMeetingControl} from '../hooks';

/**
 * WebexMeetingControl component represents an action that can
 * be taken in a meeting.
 *
 * @param {string} props.type  Name of the control as defined in adapter
 * @returns {object} JSX of the component
 */
export default function WebexMeetingControl({type}) {
  const [action, display] = useMeetingControl(type);
  const {icon, text, tooltip} = display;
  const isDisabled = display.state === MeetingControlState.DISABLED;
  const iconColor = display.state === MeetingControlState.ACTIVE ? 'red' : '';
  let button = (
    <Button color="green" size={52} ariaLabel={tooltip} onClick={action} disabled={isDisabled}>
      {text}
    </Button>
  );

  if (icon) {
    button = (
      <Button circle color={iconColor} size={56} ariaLabel={tooltip} onClick={action} disabled={isDisabled}>
        <Icon name={`${icon}_28`} />
      </Button>
    );
  }

  return button;
}

WebexMeetingControl.propTypes = {
  type: PropTypes.string.isRequired,
};
