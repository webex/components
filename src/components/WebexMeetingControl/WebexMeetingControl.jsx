import React from 'react';
import PropTypes from 'prop-types';
import {MeetingControlState} from '@webex/component-adapter-interfaces';

import webexComponentClasses from '../helpers';
import {useMeetingControl} from '../hooks';
import Button from '../generic/Button/Button';
import Icon from '../generic/Icon/Icon';
import Dropdown from '../inputs/Dropdown/Dropdown';

const controlTypeToButtonType = {
  JOIN: 'join',
  CANCEL: 'cancel',
  CLOSE: 'ghost',
  BUTTON: 'default',
  TOGGLE: 'toggle',
};

/**
 * renderButton renders a control button
 *
 * @param {Function} sc  CSS subclass function
 * @param {Function} action  Adapter control callback
 * @param {object} display  Display data of the control
 * @param {object} style  Custom style to apply
 * @param {boolean} showText  Flag that indicates whether to display text on control buttons
 * @param {boolean} asItem  Render control as an item in a list
 * @param {boolean} autoFocus  Flag indicating whether to add autofocus on a control
 * @param {number} tabIndex  TabIndex value
 * @returns {object} JSX of the component
 */
function renderButton(sc, action, display, style, showText, asItem, autoFocus, tabIndex) {
  const {
    icon,
    type,
    text,
    tooltip,
    hint,
  } = display;
  const isDisabled = display.state === MeetingControlState.DISABLED;
  const isActive = display.state === MeetingControlState.ACTIVE;

  let output;

  if (asItem) {
    output = (
      <Button
        type="ghost"
        style={style}
        size={24}
        className={sc('as-item')}
        tooltip={tooltip}
        ariaLabel={hint || text}
      >
        {icon && <Icon name={icon} size={14} className={sc('item-button-icon')} />}
        <span className={sc('item-button-text')}>{text}</span>
      </Button>
    );
  } else {
    output = (
      <Button
        className={sc('control-button')}
        type={controlTypeToButtonType[type] || 'default'}
        size={48}
        isDisabled={isDisabled}
        onClick={action}
        ariaLabel={text}
        pressed={isActive && type === 'TOGGLE'}
        tooltip={tooltip}
        autoFocus={autoFocus}
        tabIndex={tabIndex}
      >
        {icon && <Icon name={icon} size={24} />}
        {(showText || !icon) && text && <span className={sc('button-text')}>{text}</span>}
      </Button>
    );
  }

  return output;
}

/**
 * renderDropdown renders controls dropdown
 *
 * @param {Function} sc  CSS subclass function
 * @param {Function} action  Adapter control callback
 * @param {object} display  Display data of the control
 * @param {object} style  Custom style to apply
 * @param {number} tabIndex  Value of the tabIndex attribute
 * @returns {object} JSX of the component
 */
function renderDropdown(sc, action, display, style, tabIndex) {
  const {
    options, noOptionsMessage, selected, tooltip, hint,
  } = display;

  return (
    <Dropdown
      className={sc('control-select')}
      style={style}
      value={selected || ''}
      onChange={(id) => action(id)}
      options={options?.length === 0 ? [{value: '', label: noOptionsMessage}] : options}
      disabled={!options?.length}
      tooltip={tooltip}
      ariaLabel={hint}
      tabIndex={tabIndex}
    />
  );
}

/**
 * WebexMeetingControl component represents an action that can
 * be taken in a meeting.
 *
 * @param {object} props  Data passed to the component
 * @param {boolean} [props.asItem=false]  Render control as an itemin a list
 * @param {boolean} [props.autoFocus=false]  Flag indicating if the control should have autoFocus
 * @param {string} [props.className]  Custom CSS class to apply
 * @param {string} props.meetingID  ID of the meeting
 * @param {boolean} [props.showText=true]  Flag that indicates whether to display text on control buttons
 * @param {object} [props.style]  Custom style to apply
 * @param {number} props.tabIndex  Value of the tabIndex attribute
 * @param {string} props.type  Name of the control as defined in adapter
 * @returns {object} JSX of the component
 */
export default function WebexMeetingControl({
  asItem,
  autoFocus,
  className,
  meetingID,
  showText,
  style,
  tabIndex,
  type,
}) {
  const [action, display] = useMeetingControl(type, meetingID);

  const [cssClasses, sc] = webexComponentClasses(
    'meeting-control',
    className,
    {'as-item': asItem},
  );

  let output;

  if (!display || Object.keys(display).length === 0) {
    output = '';
  } else if (display.type === 'MULTISELECT') {
    output = renderDropdown(sc, action, display, style, tabIndex);
  } else {
    output = renderButton(sc, action, display, style, showText, asItem, autoFocus, tabIndex);
  }

  return <div className={cssClasses}>{output}</div>;
}

WebexMeetingControl.propTypes = {
  asItem: PropTypes.bool,
  autoFocus: PropTypes.bool,
  className: PropTypes.string,
  meetingID: PropTypes.string.isRequired,
  showText: PropTypes.bool,
  style: PropTypes.shape(),
  tabIndex: PropTypes.number,
  type: PropTypes.string.isRequired,
};

WebexMeetingControl.defaultProps = {
  asItem: false,
  autoFocus: false,
  className: '',
  showText: true,
  style: undefined,
  tabIndex: 0,
};
