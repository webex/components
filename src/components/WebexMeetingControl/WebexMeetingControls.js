import React from 'react';
import PropTypes from 'prop-types';

export const MeetingContext = React.createContext();

/**
 * WebexMeetingControls is a higher-order component that pass a meeting
 * context to several WebexMeetingControl components.
 *
 * @param {object} props
 * @returns {object} JSX of the component
 */
export default function WebexMeetingControls({meetingID, children}) {
  return (
    <MeetingContext.Provider value={meetingID}>
      <div className="meeting-controls">{children}</div>
    </MeetingContext.Provider>
  );
}

WebexMeetingControls.propTypes = {
  meetingID: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
};
