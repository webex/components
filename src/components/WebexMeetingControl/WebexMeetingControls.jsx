import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import {MeetingContext} from '../hooks';
import {WEBEX_COMPONENTS_CLASS_PREFIX} from '../../constants';

/**
 * WebexMeetingControls is a higher-order component that pass a meeting
 * context to several WebexMeetingControl components.
 *
 * @param {object} props
 * @returns {object} JSX of the component
 */
export default function WebexMeetingControls({children, className, meetingID}) {
  const mainClasses = {
    [`${WEBEX_COMPONENTS_CLASS_PREFIX}-meeting-controls`]: true,
    [className]: !!className,
  };

  return (
    <MeetingContext.Provider value={meetingID}>
      <div className={classNames(mainClasses)}>{children}</div>
    </MeetingContext.Provider>
  );
}

WebexMeetingControls.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  meetingID: PropTypes.string.isRequired,
};

WebexMeetingControls.defaultProps = {
  className: '',
};
