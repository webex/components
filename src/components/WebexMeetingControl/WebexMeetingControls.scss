.#{$WEBEX_COMPONENTS_CLASS_PREFIX}-meeting-controls {
  display: flex;
  justify-content: center;
  align-items: center;

  > * {
    margin: 0.3125rem;
  }
}
