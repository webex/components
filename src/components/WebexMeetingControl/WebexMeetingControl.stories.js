import React from 'react';
import WebexMeetingControl from './WebexMeetingControl';
import WebexMeetingControls from './WebexMeetingControls';

export default {
  title: 'Meetings/Webex Meeting Control',
  component: WebexMeetingControl,
  decorators: [(Story) => <WebexMeetingControls meetingID="meeting1"><Story /></WebexMeetingControls>],
};

const Template = (args) => <WebexMeetingControl {...args} />;

export const ActiveIcon = Template.bind({});
ActiveIcon.args = {
  type: 'mute-audio',
};

// BUG: Snapshot test fails
export const InactiveIcon = Template.bind({});
InactiveIcon.args = {
  type: 'mute-audio',
};

export const DisabledIcon = Template.bind({});
DisabledIcon.args = {
  type: 'disabled-mute-audio',
};

export const Text = Template.bind({});
Text.args = {
  type: 'join-meeting',
};

export const DisabledText = Template.bind({});
DisabledText.args = {
  type: 'disabled-join-meeting',
};
