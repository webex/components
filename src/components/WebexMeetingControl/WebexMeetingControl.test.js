import React from 'react';

import WebexMeetingControl from './WebexMeetingControl';

jest.mock('../hooks/useMeetingControl');

describe('Webex Meeting Control component', () => {
  describe('snapshot', () => {
    test('matches with text button', () => {
      expect(shallow(<WebexMeetingControl type="join-meeting" />)).toMatchSnapshot();
    });

    test('matches with disabled text button', () => {
      expect(shallow(<WebexMeetingControl type="join-meeting-disabled" />)).toMatchSnapshot();
    });

    test('matches with active icon button', () => {
      expect(shallow(<WebexMeetingControl type="mute-audio-active" />)).toMatchSnapshot();
    });

    test('matches with inactive icon button', () => {
      expect(shallow(<WebexMeetingControl type="mute-audio-inactive" />)).toMatchSnapshot();
    });

    test('matches with disabled icon button', () => {
      expect(shallow(<WebexMeetingControl type="mute-audio-disabled" />)).toMatchSnapshot();
    });
  });
});
