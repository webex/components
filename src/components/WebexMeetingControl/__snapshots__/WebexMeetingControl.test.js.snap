// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webex Meeting Control component snapshot matches with active icon button 1`] = `
<Button
  active={false}
  ariaLabel="Mute"
  ariaLabelledBy=""
  circle={true}
  className=""
  color="red"
  containerLarge={false}
  disabled={false}
  expand={false}
  href=""
  index={null}
  isButtonGroup={false}
  label=""
  large={false}
  loading={false}
  onClick={[Function]}
  removeStyle={false}
  size={56}
  style={Object {}}
  tag="button"
  type="button"
>
  <Icon
    append={false}
    ariaLabel={null}
    buttonClassName=""
    className=""
    color=""
    description=""
    isAria={true}
    name="microphone_28"
    onClick={null}
    prepend={false}
    size={null}
    sizeOverride={false}
    style={null}
    title=""
    type=""
  />
</Button>
`;

exports[`Webex Meeting Control component snapshot matches with disabled icon button 1`] = `
<Button
  active={false}
  ariaLabel="Mute disabled"
  ariaLabelledBy=""
  circle={true}
  className=""
  color=""
  containerLarge={false}
  disabled={true}
  expand={false}
  href=""
  index={null}
  isButtonGroup={false}
  label=""
  large={false}
  loading={false}
  onClick={[Function]}
  removeStyle={false}
  size={56}
  style={Object {}}
  tag="button"
  type="button"
>
  <Icon
    append={false}
    ariaLabel={null}
    buttonClassName=""
    className=""
    color=""
    description=""
    isAria={true}
    name="microphone_28"
    onClick={null}
    prepend={false}
    size={null}
    sizeOverride={false}
    style={null}
    title=""
    type=""
  />
</Button>
`;

exports[`Webex Meeting Control component snapshot matches with disabled text button 1`] = `
<Button
  active={false}
  ariaLabel="Join meeting disabled"
  ariaLabelledBy=""
  circle={false}
  className=""
  color="green"
  containerLarge={false}
  disabled={true}
  expand={false}
  href=""
  index={null}
  isButtonGroup={false}
  label=""
  large={false}
  loading={false}
  onClick={[Function]}
  removeStyle={false}
  size={52}
  style={Object {}}
  tag="button"
  type="button"
>
  Join meeting
</Button>
`;

exports[`Webex Meeting Control component snapshot matches with inactive icon button 1`] = `
<Button
  active={false}
  ariaLabel="Unmute"
  ariaLabelledBy=""
  circle={true}
  className=""
  color=""
  containerLarge={false}
  disabled={false}
  expand={false}
  href=""
  index={null}
  isButtonGroup={false}
  label=""
  large={false}
  loading={false}
  onClick={[Function]}
  removeStyle={false}
  size={56}
  style={Object {}}
  tag="button"
  type="button"
>
  <Icon
    append={false}
    ariaLabel={null}
    buttonClassName=""
    className=""
    color=""
    description=""
    isAria={true}
    name="microphone-mute_28"
    onClick={null}
    prepend={false}
    size={null}
    sizeOverride={false}
    style={null}
    title=""
    type=""
  />
</Button>
`;

exports[`Webex Meeting Control component snapshot matches with text button 1`] = `
<Button
  active={false}
  ariaLabel="Join meeting"
  ariaLabelledBy=""
  circle={false}
  className=""
  color="green"
  containerLarge={false}
  disabled={false}
  expand={false}
  href=""
  index={null}
  isButtonGroup={false}
  label=""
  large={false}
  loading={false}
  onClick={[Function]}
  removeStyle={false}
  size={52}
  style={Object {}}
  tag="button"
  type="button"
>
  Join meeting
</Button>
`;
